### PR TITLE
EZP-31521: Fixed sorting logic listener for ordered menu items

### PR DESCRIPTION
**Pull request by Nattfarinn**
| Question      | Answer
| ------------- | ---
| Tickets       | https://jira.ez.no/browse/EZP-31521 https://jira.ez.no/browse/EZEE-2953
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | yes
| Tests pass?   | yes
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

Steps to reproduce:
https://jira.ez.no/browse/EZEE-2953

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Ready for Code Review

### CODE REVIEW

File: src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        $expected = [
+            'first',
+            'second',
+            'third',
+            'fourth',
+            'fifth',
+        ];
+
+        $this->assertEquals($this->getChildrenNames($menu), $expected);
```
Comment (alongosz) [inline src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php]: per convention it's the other way around: 
```suggestion
        $this->assertEquals($expected, $this->getChildrenNames($menu));
```
(see the method prototype).

File: src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+        $expected = [
+            'second',
+            'fourth',
+            'third',
+            'fifth',
+            'first',
+        ];
+
+        $this->assertEquals($this->getChildrenNames($menu), $expected);
```
Comment (alongosz) [inline src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php]:
```suggestion
        $this->assertEquals($expected, $this->getChildrenNames($menu));
```

File: src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+        $expected = [
+            'second',
+            'third',
+            'fourth',
+            'first',
+            'fifth',
+        ];
+
+        $this->assertEquals($this->getChildrenNames($menu), $expected);
```
Comment (alongosz) [inline src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php]:
```suggestion
        $this->assertEquals($expected, $this->getChildrenNames($menu));
```

File: src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            'second',
+            'fourth',
+            'third',
+            'fifth',
+            'first',
+            'unordered',
+            'another',
+        ];
+
+        $this->assertEquals($this->getChildrenNames($menu), $expected);
```
Comment (alongosz) [inline src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php]:
```suggestion
        $this->assertEquals($expected, $this->getChildrenNames($menu));
```

File: src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+        $expectedNested = [
+            'second_child',
+            'third_child',
+            'first_child',
+            'unordered_child',
+        ];
+
+        $this->assertEquals($this->getChildrenNames($menu), $expected);
+        $this->assertEquals($this->getChildrenNames($menu->getChild('another')), $expectedNested);
```
Comment (alongosz) [inline src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php]:
```suggestion
        $this->assertEquals($expected, $this->getChildrenNames($menu));
        $this->assertEquals($expectedNested, $this->getChildrenNames($menu->getChild('another')));
```

File: src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php (hunk 1 above)
Comment (Nattfarinn) [reply src/lib/Tests/Menu/ReorderByOrderNumberListenerTest.php]: @alongosz True, fixed.

**Comment by adamwojs** [comment]
This PR should be targeted to `1.5`  branch. Bug with invalid recursive `reorderMenuItems` call is present in `eZ Platform 2.5` as well.  Ref.  https://github.com/ezsystems/ezplatform-admin-ui/blob/1.5/src/lib/Menu/Listener/ReorderByOrderNumberListener.php#L27

File: src/lib/Menu/Listener/ReorderByOrderNumberListener.php [hunk 6]
```diff
@@ -9,63 +9,44 @@
...
 
-                $menuOrderArray = array_merge(
-                    array_slice($menuOrderArray, 0, $position),
-                    [$value],
-                    array_slice($menuOrderArray, $position)
-                );
+            if ($orderNumber === null) {
+                $unorderedMenuItemsList[] = $nestedMenuItem;
+            } else {
+                $menuItemList[$orderNumber][] = $nestedMenuItem;
```
Comment (webhdx) [inline src/lib/Menu/Listener/ReorderByOrderNumberListener.php]: Doesn't it raise a warning? I'd check if `$menuItemList[$orderNumber]` is not set and initialize it as an array first.

Comment (Nattfarinn) [reply src/lib/Menu/Listener/ReorderByOrderNumberListener.php]: It's correct construction. You can dynamically allocate array "on demand" this way.

**Comment by Nattfarinn** [comment]
@adamwojs Done

File: src/lib/Menu/Listener/ReorderByOrderNumberListener.php (hunk 6 above)
Comment (webhdx) [reply src/lib/Menu/Listener/ReorderByOrderNumberListener.php]: It's just too PHP specific for my taste. Anyway I'm fine with this ✅

File: src/lib/Menu/Listener/ReorderByOrderNumberListener.php [hunk 7]
```diff
@@ -9,63 +9,44 @@
...
+    private function recursiveReorderMenuItems(ItemInterface $menuItem): void
+    {
+        $menuItemList = [];
+        $unorderedMenuItemsList = [];
 
-        if (count($alreadyTaken)) {
-            foreach ($alreadyTaken as $key => $value) {
-                $keysArray = array_keys($menuOrderArray);
-                $position = array_search($key, $keysArray);
+        foreach ($menuItem->getChildren() as $key => $nestedMenuItem) {
```
Comment (adamwojs) [inline src/lib/Menu/Listener/ReorderByOrderNumberListener.php]:
```suggestion
        foreach ($menuItem->getChildren() as $nestedMenuItem) {
```

`$key` variable is unused.

Comment (adamwojs) [inline src/lib/Menu/Listener/ReorderByOrderNumberListener.php]:
```suggestion
        foreach ($menuItem->getChildren() as $nestedMenuItem) {
```

`$key` variable is unused.

**Comment by lserwatka** [comment]
@Nattfarinn could you merge it up?

**Comment by Nattfarinn** [comment]
Merged up to `master`

**Pull request merged**